plugins/orchestrator/src/utils/usePolling.ts

```diff
@@ -0,0 +1,116 @@
+import React, { useCallback, useRef } from 'react';
+import { useAsyncFn } from 'react-use';
+
+const useTimeoutFn = (fn: Function, ms: number = 0) => {
+  // couldn't use useTimeoutFn from react-use as it starts the timeout on mount
+  const timeout = useRef<ReturnType<typeof setTimeout>>();
+  const callback = useRef(fn);
+
+  const set = useCallback(() => {
+    if (timeout.current) clearTimeout(timeout.current);
+    timeout.current = setTimeout(() => {
+      callback.current();
+    }, ms);
+  }, [ms]);
+
+  const clear = useCallback(() => {
+    if (timeout.current) clearTimeout(timeout.current);
+  }, []);
+
+  // update ref when function changes
+  React.useEffect(() => {
+    callback.current = fn;
+  }, [fn]);
+
+  // clear on unmount
+  React.useEffect(() => {
+    return clear;
+  }, [clear]);
+
+  return { clear, set };
+};
+
+const usePolling = <T>(
```

Comment:
Was there any special reason you chose not to to use [useInterval](https://github.com/streamich/react-use/blob/HEAD/docs/useInterval.md) along with [useAsyncRetry](https://github.com/streamich/react-use/blob/HEAD/docs/useAsyncRetry.md) from `react-use`. I believe it could have saved you some work...

Reply:
Can we have some unit tests for this?

Reply:
useInterval doesn't support stopping and restarting the interval

Reply:
There are stories in storybook that showcase how this works
I'll add unit tests as well

Reply:
> useInterval doesn't support stopping and restarting the interval

Correct, but it does support [pausing it](https://github.com/streamich/react-use/blob/HEAD/docs/useInterval.md#:~:text=The%20interval%20can%20be%20paused%20by%20setting%20the%20delay%20to%20null).

Reply:
Right, thanks for pointing that out, I didn't notice before
Not sure how much of the code it would remove
It still has all the logic of when to stop the interval, the error counting, etc

Reply:
Example using `useAsyncRetry` and `useInterval`  

[![Edit use-polling-demo](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/p/sandbox/use-polling-demo-fzpsrf?file=%2Fsrc%2FApp.tsx%3A23%2C12)

Reply:
https://github.com/janus-idp/backstage-plugins/pull/1081/commits/f5902ffc1ab0b0c52e6567c261e788de0857f1f9